File: src/dictionary.go
package main

import "fmt"
import "bufio"
import "os"
import "strings"
import "io/ioutil"
import "sort"

import "golang.org/x/net/html"

type DictionaryOfStringArticles struct {
	Entries      []string
	CompactIndex []IndexEntry // should be sorted by words
}

type IndexEntry struct {
	word       string
	articleIds []int
}

func (d DictionaryOfStringArticles) FindCeiling(word string) int {
	return sort.Search(len(d.CompactIndex), func(i int) bool {
		fmt.Println(d.CompactIndex[i].word, word, d.CompactIndex[i].word >= word)
		return d.CompactIndex[i].word >= word
	})
}

func LoadDictionary(path string) DictionaryOfStringArticles {
	shards, err := ioutil.ReadDir(path)
	if err != nil {
		panic(err)
	}

	fmt.Printf("Loading data for dictionary %v, %v shards\n", path, len(shards))

	articles := make([]string, 0)

	for _, shard := range shards {
		file, err := os.Open(path + "/" + shard.Name())
		if err != nil {
			panic(err)
		}
		scanner := bufio.NewScanner(file)
		accumulator := make([]string, 0)
		for scanner.Scan() {
			line := scanner.Text()
			if line == "" && len(accumulator) > 0 {
				articles = append(articles, strings.Join(accumulator, " "))
				accumulator = make([]string, 0)
			} else {
				accumulator = append(accumulator, line)
			}
		}
		if len(accumulator) > 0 {
			articles = append(articles, strings.Join(accumulator, " "))
		}
	}
	index := build_index(articles)

	sort.Slice(index, func(i, j int) bool { return index[i].word < index[j].word })
	return DictionaryOfStringArticles{Entries: articles, CompactIndex: index}
}

func positionToWeight(index int, size int) float64 {
	return 1.0 - (float64(index) / float64(size))
}

type Weighted struct {
	index  int
	weight float64
}

func build_index(articles []string) []IndexEntry {
	index_with_weights := map[string][]Weighted{}

	for article_index, entry := range articles {
		doc := html.NewTokenizer(strings.NewReader(entry))
		keywords := []string{}

		iterating := true
		for iterating {
			tt := doc.Next()
			switch {
			case tt == html.ErrorToken:
				iterating = false
			case tt == html.StartTagToken:
				t := doc.Token()
				for _, a := range t.Attr {
					if a.Key == "dict-key" {
						keywords = append(keywords, strings.Split(a.Val, ",")...)
						break
					}
				}
			}
		}

		for kw_index, kw := range keywords {
			weight := positionToWeight(kw_index, len(keywords))
			entry, exists := index_with_weights[kw]
			if !exists {
				entry = []Weighted{}
			}
			index_with_weights[kw] = append(entry, Weighted{article_index, weight})
		}
	}

	compact_index := make([]IndexEntry, 0, len(articles))

	for key, weighted_entry_indices := range index_with_weights {
		// reverse sort
		sort.Slice(weighted_entry_indices,
			func(i, j int) bool { return weighted_entry_indices[j].weight < weighted_entry_indices[i].weight })
		indices := []int{}
		for _, w := range weighted_entry_indices {
			indices = append(indices, w.index)
		}
		compact_index = append(compact_index, IndexEntry{strings.ToLower(key), indices})
	}

	return compact_index
}

File: src/dictionary.rs
extern crate select;

extern crate serde;
extern crate serde_cbor;

use std::collections::BTreeMap;
use std::fs;
use std::fs::File;
use std::io::BufReader;
use std::io::BufRead;
use std::path::PathBuf;
use std::time::SystemTime;

use self::select::document::Document;
use self::select::predicate::Attr;

#[derive(Serialize, Deserialize, Debug)]
pub struct DictionaryOfStringArticles {
    pub entries: Vec<String>,
    pub compact_index: BTreeMap<String, Vec<usize>>
}

fn shards(path: &str) -> Vec<PathBuf> {
    fs::read_dir(path).unwrap().map(|x|x.unwrap().path()).filter(|x|x.extension().unwrap().to_str() == Some("txt")).collect()
}

fn shards_latest_timestamp(path: &str) -> SystemTime {
    shards(path).iter().map(|x|fs::metadata(x).unwrap().modified().unwrap()).max().unwrap()
}

pub fn load_dictionary(path: String) -> DictionaryOfStringArticles {
    let all_cbor = path.clone() + ".cached.cbor";
    let maybe_metadata = fs::metadata(&all_cbor);
    let needs_update = match maybe_metadata {
        Ok(meta) => meta.modified().unwrap() < shards_latest_timestamp(&path),
        Err(_) => true
    };
    if needs_update {
        let parsed = parse_from_txt(path);
        serde_cbor::to_writer(&mut File::create(all_cbor).unwrap(), &parsed).unwrap();
        return parsed;
    } else {
        println!("Loading data for dictionary {}, from pre-parsed CBOR", path);
        return serde_cbor::from_reader(File::open(all_cbor).unwrap()).unwrap();
    }
}

pub fn parse_from_txt(path: String) -> DictionaryOfStringArticles {
    let mut shards: Vec<PathBuf> = shards(&path);
    shards.sort();
    println!("Loading data for dictionary {}, from {} text shards", path, shards.len());

    let mut articles: Vec<String> = vec![];

    for shard in shards {
        let mut f = BufReader::new(File::open(shard).unwrap());
        let mut accumulator = String::new();
        for line_res in f.lines() {
            let line = line_res.unwrap();
            if line == "" && accumulator != "" {
                articles.push(accumulator);
                accumulator = String::new();
            } else {
                if accumulator != "" {
                    accumulator.push_str(" ");
                }
                accumulator.push_str(&line);
            }
        }
        if accumulator != "" {
            articles.push(accumulator);
        }
    }
    let index = build_index(&articles);
    return DictionaryOfStringArticles{entries: articles, compact_index: index}
}

fn position_to_weight(index: usize, size: usize) -> f64 {
    return 1.0 - (index as f64 / size as f64)
}

fn build_index(articles: &Vec<String>) -> BTreeMap<String, Vec<usize>> {
    let mut index_with_weights: BTreeMap<String, Vec<(usize, f64)>> = BTreeMap::new();

    for (article_index, entry) in articles.iter().enumerate() {
        let html = Document::from(&entry as &str);
        let keywords: Vec<&str> = 
            html.find(Attr("dict-key", ())).flat_map(|node| node.attr("dict-key").unwrap().split(',')).collect();

        for (kw_index, &kw) in keywords.iter().enumerate() {
            let weight = position_to_weight(kw_index, keywords.len());
            index_with_weights.entry(kw.to_owned()).or_insert_with(|| Vec::new()).push((article_index, weight));
        }
    }
    let mut compact_index: BTreeMap<String, Vec<usize>> = BTreeMap::new();
    for (key, weighted_entry_indices) in index_with_weights.iter_mut() {
        weighted_entry_indices.sort_by(|a,b|a.partial_cmp(b).unwrap());
        compact_index.insert(key.to_lowercase(), weighted_entry_indices.iter().map(|x|x.0).collect());
    }

    return compact_index
}